src/pipecat/services/ultravox/llm.py

```diff
@@ -0,0 +1,534 @@
+            else:
+                await self._socket.send(json.dumps(content))
+        except Exception as e:
+            if self._disconnecting or not self._socket:
+                return
+            await self.push_error("Ultravox websocket send error", e, fatal=True)
+
+    #
+    # response handling
+    #
+
+    async def _receive_messages(self):
+        """Receive messages from the Ultravox Realtime WebSocket."""
+        async for message in self._socket:
+            try:
+                if isinstance(message, bytes):
+                    await self._handle_audio(message)
+                    continue
+
+                data = json.loads(message)
+                match data.get("type"):
+                    case "state":
+                        if self._bot_responding and data.get("state") != "speaking":
+                            await self._handle_response_end()
+                    case "client_tool_invocation":
+                        await self._handle_tool_invocation(
+                            data.get("toolName"), data.get("invocationId"), data.get("parameters")
+                        )
+                    case "transcript":
+                        match data.get("role"):
+                            case "user":
+                                if not data.get("final"):
+                                    logger.warning(
+                                        "Unexpected non-final user transcript from Ultravox Realtime; ignoring."
+                                    )
+                                else:
+                                    await self._handle_user_transcript(data.get("text"))
+                            case "agent":
+                                await self._handle_agent_transcript(
+                                    data.get("medium"),
+                                    data.get("text"),
+                                    data.get("delta"),
+                                    data.get("final", False),
+                                )
+                            case _:
+                                logger.debug(
+                                    f"Received transcript with unknown role from Ultravox Realtime: {data}"
+                                )
+                    case _:
+                        logger.debug(f"Received unhandled Ultravox message: {data}")
+            except Exception as e:
+                if self._disconnecting or not self._socket:
+                    return
+                await self.push_error("Ultravox websocket receive error", e, fatal=True)
+
+    async def _handle_audio(self, audio: bytes):
+        """Handle incoming audio bytes from Ultravox Realtime."""
+        if not audio:
+            return
+        if not self._bot_responding:
+            await self.push_frame(BotStartedSpeakingFrame())
+            await self.push_frame(TTSStartedFrame())
+            await self.push_frame(LLMFullResponseStartFrame())
```

Comment:
- No need to push `BotStartedSpeakngFrame` this is automatically push by the output transport when output audio frames get there.
- Push `LLMFullResponseStartFrame` before `TTSStartedFrame`. It should be fine, but usually that's what happens.

So
```
            await self.push_frame(LLMFullResponseStartFrame())
            await self.push_frame(TTSStartedFrame())
```